internal/datanode/importv2/util.go

```diff
@@ -207,6 +199,156 @@ func AppendSystemFieldsData(task *ImportTask, data *storage.InsertData, rowNum i
+			continue
+		}
+		if tempData, ok := data.Data[field.GetFieldID()]; ok {
+			if tempData.RowNum() > 0 {
+				continue // values have been read from data file
+			}
+		}
+
+		// add a new column and fill with null or default
+		dataType := field.GetDataType()
+		fieldData, err := storage.NewFieldData(dataType, field, rowNum)
+		if err != nil {
+			return err
+		}
+		data.Data[field.GetFieldID()] = fieldData
+
+		nullable := field.GetNullable()
+		defaultVal := field.GetDefaultValue()
+
+		// bool/int8/int16/int32/int64/float/double/varchar/json/array can be null value
+		// bool/int8/int16/int32/int64/float/double/varchar can be default value
+		switch dataType {
+		case schemapb.DataType_Bool:
+			appender := &nullDefaultAppender[bool]{}
+			if defaultVal != nil {
+				v := defaultVal.GetBoolData()
+				err = appender.AppendDefault(fieldData, v, rowNum)
+			} else if nullable {
+				err = appender.AppendNull(fieldData, rowNum)
+			}
+		case schemapb.DataType_Int8:
+			appender := &nullDefaultAppender[int8]{}
+			if defaultVal != nil {
+				v := defaultVal.GetIntData()
+				err = appender.AppendDefault(fieldData, int8(v), rowNum)
+			} else if nullable {
+				err = appender.AppendNull(fieldData, rowNum)
+			}
+		case schemapb.DataType_Int16:
+			appender := &nullDefaultAppender[int16]{}
+			if defaultVal != nil {
+				v := defaultVal.GetIntData()
+				err = appender.AppendDefault(fieldData, int16(v), rowNum)
+			} else if nullable {
+				err = appender.AppendNull(fieldData, rowNum)
+			}
+		case schemapb.DataType_Int32:
+			appender := &nullDefaultAppender[int32]{}
+			if defaultVal != nil {
+				v := defaultVal.GetIntData()
+				err = appender.AppendDefault(fieldData, int32(v), rowNum)
+			} else if nullable {
+				err = appender.AppendNull(fieldData, rowNum)
+			}
+		case schemapb.DataType_Int64:
+			appender := &nullDefaultAppender[int64]{}
+			if defaultVal != nil {
+				v := defaultVal.GetLongData()
+				err = appender.AppendDefault(fieldData, v, rowNum)
+			} else if nullable {
+				err = appender.AppendNull(fieldData, rowNum)
+			}
+		case schemapb.DataType_Float:
+			appender := &nullDefaultAppender[float32]{}
+			if defaultVal != nil {
+				v := defaultVal.GetFloatData()
+				err = appender.AppendDefault(fieldData, v, rowNum)
+			} else if nullable {
+				err = appender.AppendNull(fieldData, rowNum)
+			}
+		case schemapb.DataType_Double:
+			appender := &nullDefaultAppender[float64]{}
+			if defaultVal != nil {
+				v := defaultVal.GetDoubleData()
+				err = appender.AppendDefault(fieldData, v, rowNum)
+			} else if nullable {
+				err = appender.AppendNull(fieldData, rowNum)
+			}
+		case schemapb.DataType_VarChar:
+			appender := &nullDefaultAppender[string]{}
+			if defaultVal != nil {
+				v := defaultVal.GetStringData()
+				err = appender.AppendDefault(fieldData, v, rowNum)
+			} else if nullable {
+				err = appender.AppendNull(fieldData, rowNum)
+			}
+		case schemapb.DataType_JSON:
+			if nullable {
```

Comment:
do we need to return error if defaultVal is not nil?

Reply:
Now milvus doesn't allow to json/array field with a default value, no need to check "defaultVal is not nil" here.